### PR TITLE
feature: add support of MacVtap

### PR DESCRIPTION
**Pull request by ffmancera**
Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>

### CODE REVIEW

**Comment by ffmancera** [comment]
For some reason, this is not working for me. The "MacVtap" type is being recognised but not matched. Not sure why and I will continue investigating this.

Current output:

```
Unhandled iface type Other("Tun")
Unhandled iface type Other("MacVtap")
{
  "name": "mac0",
  "iface_type": {
    "Other": "MacVtap"
  },
  "state": "Up",
  "mtu": 1500,
  "flags": [
    "Broadcast",
    "LowerUp",
    "Multicast",
    "Running",
    "Up"
  ],
  "ipv6": {
    "addresses": [
      {
        "address": "fe80::5075:beff:fe02:675d",
        "prefix_len": 64,
        "valid_lft": "forever",
        "preferred_lft": "forever"
      }
    ]
  },
  "mac_address": "52:75:BE:02:67:5D"
}
```

Could you take a look? Thanks, @cathay4t.

**Review by cathay4t — CHANGES_REQUESTED** [review]
Too much duplicate code against mac vlan. Please find a way to dedup.

**Comment by ffmancera** [comment]
> Too much duplicate code against mac vlan. Please find a way to dedup.

Done. Now mac_v.rs is handling mac_vtap and mac_vlan interfaces. This way there is no duplicated code. What do you think? Thanks!

File: tools/test_env [hunk 1]
```diff
@@ -84,6 +84,13 @@ elif [ "CHK$1" == "CHKmacvlan" ];then
     sudo ip link set link dev mac0 type macvlan macaddr add 00:22:22:22:22:22
     sudo ip link set eth1 up
     sudo ip link set mac0 up
+elif [ "CHK$1" == "CHKmacvtap" ];then
+    create_nics
+    sudo ip link add macvtap0 link eth1 type macvtap mode source
+    sudo ip link set link dev macvtap0 type macvtap macaddr add 00:11:11:11:11:11
+    sudo ip link set link dev macvtap0 type macvtap macaddr add 00:22:22:22:22:22
+    sudo ip link set eth1 up
+    sudo ip link set macvtap0 up
```
Comment (cathay4t) [inline tools/test_env:93]: You forgot the delete this interface in `rm` section.

File: src/lib/ifaces/mac_v.rs [hunk 2]
```diff
@@ -8,24 +8,24 @@ use netlink_packet_route::rtnl::nlas::NlasIterator;
...
-const IFLA_MACVLAN_FLAGS: u16 = 2;
+pub const IFLA_MACV_MODE: u16 = 1;
+pub const IFLA_MACV_FLAGS: u16 = 2;
 // const IFLA_MACVLAN_MACADDR_MODE: u16 = 3;    // not used in query
-const IFLA_MACVLAN_MACADDR: u16 = 4;
-const IFLA_MACVLAN_MACADDR_DATA: u16 = 5;
-const IFLA_MACVLAN_MACADDR_COUNT: u16 = 6;
+pub const IFLA_MACV_MACADDR: u16 = 4;
+pub const IFLA_MACV_MACADDR_DATA: u16 = 5;
+pub const IFLA_MACV_MACADDR_COUNT: u16 = 6;
```
Comment (cathay4t) [inline src/lib/ifaces/mac_v.rs]: There is no need for use to expose these constants. Please remove the `pub`.

File: src/lib/ifaces/mac_v.rs [hunk 3]
```diff
@@ -8,24 +8,24 @@ use netlink_packet_route::rtnl::nlas::NlasIterator;
...
-const MACVLAN_MODE_PRIVATE: u32 = 1;
-const MACVLAN_MODE_VEPA: u32 = 2;
-const MACVLAN_MODE_BRIDGE: u32 = 4;
-const MACVLAN_MODE_PASSTHRU: u32 = 8;
-const MACVLAN_MODE_SOURCE: u32 = 16;
+pub const MACV_MODE_PRIVATE: u32 = 1;
+pub const MACV_MODE_VEPA: u32 = 2;
+pub const MACV_MODE_BRIDGE: u32 = 4;
+pub const MACV_MODE_PASSTHRU: u32 = 8;
+pub const MACV_MODE_SOURCE: u32 = 16;
```
Comment (cathay4t) [inline src/lib/ifaces/mac_v.rs]: These constant is copy from kernel, don't change the name of it. Just add a comment says the `mac_vlan` are sharing the same code with `mac_vtap` in kernel for netlink informations.

**Review by cathay4t — CHANGES_REQUESTED** [review]
The `MacVInfo` naming is against Rust API guideline.

And I dislike the name `MacV`.
For public interface, user don't need to know what is `MacV` is and why it should be.

User should still get `MacVlanInfo` and `MacVtapInfo`.

Implement a simple `from MacVlanInfo -> MacVtapInfo` in `src/lib/ifaces/mac_vtap.rs` could deduplicate the code with `mac_vlan.rs`.

File: tools/test_env (hunk 1 above)
Comment (ffmancera) [reply tools/test_env:93]: Oops. Fixed, thanks!

File: src/lib/ifaces/mac_v.rs (hunk 2 above)
Comment (ffmancera) [reply src/lib/ifaces/mac_v.rs]: Done, thanks!

**Comment by ffmancera** [comment]
> The `MacVInfo` naming is against Rust API guideline.
> 
> And I dislike the name `MacV`.
> For public interface, user don't need to know what is `MacV` is and why it should be.
> 
> User should still get `MacVlanInfo` and `MacVtapInfo`.
> 
> Implement a simple `from MacVlanInfo -> MacVtapInfo` in `src/lib/ifaces/mac_vtap.rs` could deduplicate the code with `mac_vlan.rs`.

Done, thanks!

File: src/lib/ifaces/mac_vtap.rs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    Unknown,
+}
+
+impl Default for MacVtapMode {
+    fn default() -> Self {
+        MacVtapMode::Unknown
+    }
+}
+
+impl From<u32> for MacVtapMode {
```
Comment (cathay4t) [inline src/lib/ifaces/mac_vtap.rs]: This is not required and the constants it used.

**Review by cathay4t — CHANGES_REQUESTED** [review]

File: src/lib/ifaces/mac_vtap.rs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+            flags: d.flags,
+            allowed_mac_addresses: d.allowed_mac_addresses,
+        }
+    }
+
+}
+pub(crate) fn get_mac_vtap_info(data: &nlas::InfoData) -> Option<MacVtapInfo> {
+    if let Some(info) = get_mac_vlan_info(data) {
+        let mac_vtap_info = MacVtapInfo::from(info);
+        Some(mac_vtap_info)
```
Comment (cathay4t) [inline src/lib/ifaces/mac_vtap.rs]: `Some(info.into())`

**Review by cathay4t — CHANGES_REQUESTED** [review]

File: src/lib/ifaces/mac_vtap.rs (hunk 4 above)
Comment (ffmancera) [reply src/lib/ifaces/mac_vtap.rs]: Done, thanks!

File: src/lib/ifaces/mac_vtap.rs (hunk 5 above)
Comment (ffmancera) [reply src/lib/ifaces/mac_vtap.rs]: Done, thanks!

**Review by cathay4t — APPROVED** [review]

**Pull request merged**